### PR TITLE
[TIR] Add PreOrderVisit and VisitPrimFuncs

### PR DESCRIPTION
**Pull request by junrushao1994**
Another prerequisite for PR M1b on TensorIR upstreaming work (#7527)

### CODE REVIEW

File: include/tvm/tir/stmt_functor.h [hunk 1]
```diff
@@ -386,6 +394,24 @@ inline T Substitute(T input, const std::unordered_map<const VarNode*, PrimExpr>&
...
+ */
+TVM_DLL void PreOrderVisit(const ObjectRef& node,
+                           const std::function<bool(const ObjectRef&)>& fvisit);
+
+/*!
+ * \brief Visit the PrimFuncs in the IRModule
+ * \param mod The IRModule to be visited
+ * \param fvisit The visitor to the PrimFuncs in the IRModule
+ */
+TVM_DLL void VisitPrimFuncs(const IRModule& mod,
```
Comment (tqchen) [inline include/tvm/tir/stmt_functor.h:412]: Consider make an inline function so we can templatize fvisit, and let compiler inline that lambda

File: include/tvm/tir/stmt_functor.h [hunk 2]
```diff
@@ -386,6 +394,24 @@ inline T Substitute(T input, const std::unordered_map<const VarNode*, PrimExpr>&
   return Substitute(std::move(input), vmap);
 }
 
+/*!
+ * \brief Recursively visit the IR in pre DFS order node, apply fvisit.
+ * If fvisit returns false, it stops visit the children of the current node.
+ * \param node The IR to be visited, can be either Stmt or Expr.
+ * \param fvisit The visitor function to be applied. If fvisit returns false, it stops visit the
```
Comment (tqchen) [inline include/tvm/tir/stmt_functor.h:401]: it won't visit the children of the node

File: include/tvm/tir/stmt_functor.h [hunk 3]
```diff
@@ -36,7 +36,15 @@
 #include <utility>
 
 namespace tvm {
+
+// Forward declaration
+class IRModule;
```
Comment (tqchen) [inline include/tvm/tir/stmt_functor.h:41]: simply include ir/module.h

File: include/tvm/tir/stmt_functor.h (hunk 1 above)
Comment (tqchen) [reply include/tvm/tir/stmt_functor.h:412]: Consider move to https://github.com/apache/tvm/blob/main/include/tvm/tir/analysis.h

File: include/tvm/tir/stmt_functor.h [hunk 4]
```diff
@@ -386,6 +394,24 @@ inline T Substitute(T input, const std::unordered_map<const VarNode*, PrimExpr>&
...
 }
 
+/*!
+ * \brief Recursively visit the IR in pre DFS order node, apply fvisit.
+ * If fvisit returns false, it stops visit the children of the current node.
+ * \param node The IR to be visited, can be either Stmt or Expr.
+ * \param fvisit The visitor function to be applied. If fvisit returns false, it stops visit the
+ * children on that node.
+ */
+TVM_DLL void PreOrderVisit(const ObjectRef& node,
```
Comment (tqchen) [inline include/tvm/tir/stmt_functor.h:404]: node=> stmt_or_expr

File: include/tvm/tir/stmt_functor.h [hunk 5]
```diff
@@ -386,6 +394,24 @@ inline T Substitute(T input, const std::unordered_map<const VarNode*, PrimExpr>&
   return Substitute(std::move(input), vmap);
 }
 
+/*!
+ * \brief Recursively visit the IR in pre DFS order node, apply fvisit.
+ * If fvisit returns false, it stops visit the children of the current node.
+ * \param node The IR to be visited, can be either Stmt or Expr.
+ * \param fvisit The visitor function to be applied. If fvisit returns false, it stops visit the
+ * children on that node.
+ */
```
Comment (tqchen) [inline include/tvm/tir/stmt_functor.h:403]: please add a testcase to https://github.com/apache/tvm/blob/main/tests/cpp/ir_functor_test.cc

File: include/tvm/tir/stmt_functor.h (hunk 1 above)
Comment (junrushao1994) [reply include/tvm/tir/stmt_functor.h:412]: Sounds good. I am going to move it to analysis.h and change it from std::function to FLambda

**Review by tqchen — APPROVED** [review]

**Review by jcf94 — APPROVED** [review]

**Pull request merged**